### PR TITLE
DateTimeType: new methods toZone and toLocaleZone

### PR DESCRIPTION
**Pull request by lolodomo**
Enhanced tests

Signed-off-by: Laurent Garnier <lg.hc@free.fr>

### CODE REVIEW

**Comment by lolodomo** [comment]
@cweitkamp : here are the additional format methods with tests. This works well and as expected.

Now, we have the choice to link the old format methods with the new format methods using  ZoneId.systemDefault() as zone. This will impact all calls to the old format methods. Impact not yet evaluated.
Or we keep unchanged the existing format methods that will continue ignoring the system timezone for formatting dates and I just update the UI code to use the new format methods with ZoneId.systemDefault() as parameter to get dates in local timezone in UI.

File: bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DateTimeTypeTest.java [hunk 1]
```diff
@@ -231,25 +231,71 @@ public void parsingTest() {
...
 
         DateTimeType dt2 = new DateTimeType(dt1.getZonedDateTime().withZoneSameInstant(ZoneId.of("Europe/Berlin")));
         assertThat(dt2, not(is(dt1)));
         assertThat(dt2.toString(), is("2019-05-25T19:53:10.000+0200"));
 
         assertThat(dt2.format(null), is("2019-05-25T19:53:10"));
         assertThat(dt2.format("%1$td.%1$tm.%1$tY %1$tH:%1$tM"), is("25.05.2019 19:53"));
-
-        assertThat(dt2.format(Locale.GERMAN, "%1$td.%1$tm.%1$tY %1$tH:%1$tM"), is("25.05.2019 19:53"));
+        assertThat(dt1.format(null, gmt), is("2019-05-25T17:53:10"));
```
Comment (cweitkamp) [inline bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DateTimeTypeTest.java]: Should be `dt2` in this and the following lines.

**Comment by lolodomo** [comment]
I am going to enhance the tests.

Ok for the idea of the toZone method rather than my new format methods.

**Comment by lolodomo** [comment]
@cweitkamp : PR is now finished and ready for a review.

I will change the REST sitemap by using the new methods in a following PR.

File: bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/library/types/DateTimeType.java [hunk 2]
```diff
@@ -108,7 +109,7 @@ public DateTimeType(String zonedValue) {
             throw new IllegalArgumentException(zonedValue + " is not in a valid format.", invalidFormatException);
         }
 
-        zonedDateTime = date;
+        zonedDateTime = date.withFixedOffsetZone();
```
Comment (cweitkamp) [inline bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/library/types/DateTimeType.java]: It looks like the right thing to streamline the SPI. But in opposition I found https://github.com/eclipse/smarthome/issues/5076 which questions if `withFixedOffsetZone()` is the right way to go. Unfortunately with no result. @openhab/core-maintainers wdyt?

Comment (lolodomo) [reply bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/library/types/DateTimeType.java]: withFixedOffsetZone() is applied to the other constructors so it makes sense to apply to this one too. It allows to get identical DateTimeType when calling the different constructors with the same data.

File: bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/library/types/DateTimeType.java [hunk 3]
```diff
@@ -140,6 +141,22 @@ public String format(Locale locale, String pattern) {
...
+        toZone(ZoneId.systemDefault());
+    }
+
+    /**
+     * Translate to a given zone
+     *
+     * @param zone the destination zone
+     */
+    public void toZone(ZoneId zone) {
+        zonedDateTime = zonedDateTime.withZoneSameInstant(zone);
```
Comment (cweitkamp) [inline bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/library/types/DateTimeType.java]: You might consider to return a new copy of DateTimeType here.

**Comment by lolodomo** [comment]
Ok.  I hesitated between the two options.

File: bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/library/types/DateTimeType.java [hunk 4]
```diff
@@ -140,6 +141,22 @@ public String format(Locale locale, String pattern) {
...
+     * Translate to the locale time zone
+     */
+    public void toLocaleZone() {
+        toZone(ZoneId.systemDefault());
+    }
+
+    /**
+     * Translate to a given zone
+     *
+     * @param zone the destination zone
```
[the remark this answers is not in the document]
Comment (lolodomo) [reply bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/library/types/DateTimeType.java]: Ok I will add it.

**Comment by lolodomo** [comment]
@cweitkamp : changes done.
Please let me know if you prefer that the methods throw an exception or return null in case of problem.

File: bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DateTimeTypeTest.java [hunk 5]
```diff
@@ -232,79 +314,105 @@ public void parsingTest() {
...
+    public void changingZoneTest() {
+        DateTimeType dt = createDateTimeType();
+        DateTimeType dt2 = dt.toLocaleZone();
+        assertEquals(parameterSet.expectedResultLocalTZ, dt2.toFullString());
+        dt2 = dt.toZone(parameterSet.defaultTimeZone.toZoneId());
+        assertEquals(parameterSet.expectedResultLocalTZ, dt2.toFullString());
+        boolean thrown = false;
+        try {
+            dt2 = dt.toZone("XXX");
+        } catch (DateTimeException e) {
```
Comment (cweitkamp) [inline bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DateTimeTypeTest.java]: This should be an own test which expects the thrown exception:

```
    @Test(expected = DateTimeException.class)
    public void changingZoneThrowsExceptionTest() {
        // TODO
    }
```

File: bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DateTimeTypeTest.java [hunk 6]
```diff
@@ -104,13 +136,48 @@ public ParameterSet(TimeZone defaultTimeZone, Map<String, Integer> inputTimeMap,
...
+         * @param inputTimeMap
+         * @param inputTimeZone
+         * @param inputTimeString
+         * @param expectedResult
+         * @param expectedResultLocalTZ
+         * @param locale
+         * @param pattern
+         * @param expectedFormattedResult
+         */
+        public ParameterSet(TimeZone defaultTimeZone, Map<String, Integer> inputTimeMap, TimeZone inputTimeZone,
```
Comment (cweitkamp) [inline bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DateTimeTypeTest.java]: I am just wondering if we can reuse this constructor called by the other constructors. But that is only a side comment. Up to you.

File: bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DateTimeTypeTest.java [hunk 7]
```diff
@@ -38,6 +39,7 @@
  * @author Thomas.Eichstaedt-Engelen - Initial contribution
  * @author Gaël L'hopital - Added Timezone and Milliseconds
  * @author Erdoan Hadzhiyusein - Added ZonedDateTime tests
+ * @author Laurent Garnier - Enhanced tests
  */
 @RunWith(Parameterized.class)
 public class DateTimeTypeTest {
```
Comment (cweitkamp) [inline bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DateTimeTypeTest.java]: Looking at the properties of the inner class `ParameterSet` does it make sense to add nullness annotations here?

**Comment by lolodomo** [comment]
Do you know if this is possible to run a particular test only once and not for each ParsmeterSet?

File: bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DateTimeTypeTest.java (hunk 6 above)
Comment (lolodomo) [reply bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DateTimeTypeTest.java]: Yes we can.

File: bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DateTimeTypeTest.java (hunk 5 above)
Comment (lolodomo) [reply bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DateTimeTypeTest.java]: Done

File: bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DateTimeTypeTest.java (hunk 6 above)
Comment (lolodomo) [reply bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DateTimeTypeTest.java]: Done

File: bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DateTimeTypeTest.java (hunk 7 above)
Comment (lolodomo) [reply bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DateTimeTypeTest.java]: Done

**Comment by cweitkamp** [comment]
No, I am afraid, I do not know. But does it hurt?

**Comment by lolodomo** [comment]
No, it just certainly takes few additional ms to run.

**Comment by kaikreuzer** [comment]
@cweitkamp I am honestly not really clear on the effect of that change, but as you have looked deeper into it and approved it, I don't see any reason not to merge it :-)

**Comment by lolodomo** [comment]
Cool, I can move to the next step, using these methods to fix a bug.

**Comment by cweitkamp** [comment]
Just for clarification: From https://docs.oracle.com/javase/8/docs/api/java/time/ZonedDateTime.html#withFixedOffsetZone--

> This returns a zoned date-time where the zone ID is the same as getOffset(). The local date-time, offset and instant of the result will be the same as in this date-time.
>
> Setting the date-time to a fixed single offset means that any future calculations, such as addition or subtraction, have no complex edge cases due to time-zone rules. This might also be useful when sending a zoned date-time across a network, as most protocols, such as ISO-8601, only handle offsets, and not region-based zone IDs.

Which basically means that it cuts off the "US/Central" (or whatever timezone you are in) string from the object (you can see that in the debugger). The quoted text above also states that cutting it off simplifies calculations based on this object later on (e.g. `2018-11-04T01:25:43-06:00[US/Central]` results in `2018-11-04T01:25:43-06:00`).

**Pull request merged**